src/config.rs

```diff
@@ -32,34 +34,93 @@ impl FromStr for WorkMode {
 }
 
 #[derive(Debug)]
-pub enum ConfigError {
-    Io(IoError, String),
-    TomlError(String, String),
+pub enum Error {
+    CommandLine(DocoptError),
+    Io(IoError, PathBuf),
+    Toml(String, PathBuf),
+    Options(String, PathBuf),
 }
 
-impl Error for ConfigError {
+impl StdError for Error {
     fn description(&self) -> &str {
+        use self::Error::*;
         match *self {
-            ConfigError::Io(ref e, _) => e.description(),
-            ConfigError::TomlError(ref s, _) => s,
+            CommandLine(ref e) => e.description(),
+            Io(ref e, _) => e.description(),
+            Toml(ref s, _) => s,
+            Options(ref s, _) => s,
         }
     }
 }
 
-impl Display for ConfigError {
+impl Display for Error {
     fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        use self::Error::*;
         match *self {
-            ConfigError::Io(ref err, ref filename) => {
-                try!(write!(f, "\"{}\": ", filename));
+            CommandLine(ref err) => err.fmt(f),
+            Io(ref err, ref filename) => {
+                try!(write!(f, "Failed to read config \"{}\": ", filename.display()));
                 err.fmt(f)
             }
-            ConfigError::TomlError(ref err, ref filename) => {
-                write!(f, "\"{}\": {}", filename, err)
+            Toml(ref err, ref filename) => {
+                write!(f, "\"{}\": {}", filename.display(), err)
+            }
+            Options(ref err, ref filename) => {
+                write!(f, "\"{}\": {}", filename.display(), err)
             }
         }
     }
 }
 
+impl<'a> From<DocoptError> for Error {
+    fn from(e: DocoptError) -> Error {
+        Error::CommandLine(e)
+    }
+}
+
+impl<P: AsRef<Path>> From<(IoError, P)> for Error {
+    fn from(e: (IoError, P)) -> Error {
+        Error::Io(e.0, e.1.as_ref().into())
```

Comment:
`as_ref().into()` was bothering me so I come up with this:
```rust
impl<P: AsRef<OsStr> + AsRef<Path>> From<(IoError, P)> for Error {
    fn from(e: (IoError, P)) -> Error {
        Error::Io(e.0, PathBuf::from(&e.1))
    }
}
```
The downside is having to require `P: AsRef<OsStr> + AsRef<Path>` everywhere but it should be harmless because `Path` and `OsStr` essentially are interchangeable.